macros/src/buffer.rs

```diff
@@ -0,0 +1,175 @@
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote};
+use syn::{parse_quote, Ident, ItemStruct, Type};
+
+use crate::Result;
+
+pub(crate) fn impl_joined_value(input_struct: &ItemStruct) -> Result<TokenStream> {
+    let struct_ident = &input_struct.ident;
+    let (impl_generics, ty_generics, where_clause) = input_struct.generics.split_for_impl();
+    let (field_ident, field_type) = get_fields_map(&input_struct.fields)?;
+    let BufferStructConfig {
+        struct_name: buffer_struct_ident,
+    } = BufferStructConfig::from_data_struct(&input_struct);
+    let buffer_struct_vis = &input_struct.vis;
+
+    let buffer_struct: ItemStruct = parse_quote! {
+        #[derive(Clone)]
+        #[allow(non_camel_case_types)]
+        #buffer_struct_vis struct #buffer_struct_ident #impl_generics #where_clause {
+            #(
+                #buffer_struct_vis #field_ident: ::bevy_impulse::Buffer<#field_type>,
+            )*
+        }
+    };
+
+    let impl_buffer_map_layout = impl_buffer_map_layout(&buffer_struct, &input_struct)?;
+    let impl_joined = impl_joined(&buffer_struct, &input_struct)?;
+
+    let gen = quote! {
+        impl #impl_generics ::bevy_impulse::JoinedValue for #struct_ident #ty_generics #where_clause {
+            type Buffers = #buffer_struct_ident #ty_generics;
+        }
+
+        #buffer_struct
+
+        impl #impl_generics #struct_ident #ty_generics #where_clause {
+            fn select_buffers(
+                builder: &mut ::bevy_impulse::Builder,
```

Comment:
I'd rather not have `builder` as an argument. I think we should leave it to users to call `output.into_buffer(builder)` from outside of `select_buffers` and have `select_buffers` just take in the specific buffer types that are expected. This will especially matter for supporting specialized buffer types like `AnyBuffer` and `JsonBuffer`.